pyvisfile/xdmf/__init__.py

```diff
@@ -0,0 +1,1119 @@
+
+        return attr
+
+
+class XdmfUnstructuredGrid(XdmfGrid):
+    def __init__(self,
+            points: DataArray,
+            connectivity: DataArray, *,
+            topology_type: Union[Topology, TopologyType],
+            name: Optional[str] = None,
+            geometry_type: Optional[GeometryType] = None):
+        if geometry_type is None:
+            geometry_type = _geometry_type_from_points(points)
+
+        if geometry_type not in (GeometryType.XY, GeometryType.XYZ):
+            # NOTE: Paraview 5.8 seems confused when using VXVY geometry types
+            raise ValueError(f"unsupported geometry type: '{geometry_type}'")
+
+        grid = Grid(parent=None, name=name)
+
+        nelements = np.prod(connectivity.shape[:-1])
+        if isinstance(topology_type, TopologyType):
+            if topology_type == TopologyType.Polyline:
+                nodes_per_element = 2
+            else:
+                nodes_per_element = None
+
+            topology = Topology(
+                    parent=grid,
+                    ttype=topology_type,
+                    nodes_per_element=nodes_per_element,
+                    number_of_elements=nelements)
+        elif isinstance(topology, Topology):
+            topology = topology_type.replace({
+                "parent": grid,
+                "number_of_elements": nelements
+                })
+        else:
+            raise TypeError(f"unsupported type: {type(topology_type).__name__}")
+
+        connectivity.as_data_item(parent=topology)
+
+        geometry = Geometry(parent=grid, gtype=geometry_type)
+        points.as_data_item(parent=geometry)
+
+        super().__init__(grid)
+
+
+class XdmfStructuredGrid(XdmfGrid):
+    def __init__(self):
+        super().__init__()
+
+# }}}
+
+
+# {{{ writer
+
+class XdmfWriter(ElementTree):
+    def __init__(self,
+            grids: Tuple[XdmfGrid, ...], *,
+            arrays: Optional[Tuple[DataArray, ...]] = None,
+            tags: Optional[Tuple[XdmfElement, ...]] = None):
+        r"""
+        :param grids: a :class:`tuple` of grids to be added to the
+            top :class:`Domain`.
+        :param arrays: additional :class:`DataArray`\ s to be added to the
+            top :class:`Domain`, as opposed to as attribute on the grids.
+        """
+        root = Element("Xdmf", {
+            "xmlns:xi": "http://www.w3.org/2001/XInclude",
+            "Version": "3.0",
+            })
+
+        domain = Domain(parent=root)
+        if arrays is not None:
+            for ary in arrays:
+                ary.as_data_item(parent=domain)
+
+        if tags is not None:
+            for tag in tags:
+                domain.append(tag)
+
+        for grid in grids:
+            domain.append(grid.getroot())
+
+        super().__init__(root)
+
+    def write_pretty(self, filename):
+        from xml.etree.ElementTree import tostring
+        from xml.dom import minidom
+        dom = minidom.parseString(tostring(
```

Comment:
How come this needs to go through `minidom`? (At least add a comment explaining why)

Reply:
Mostly just copied from [stackoverflow](https://stackoverflow.com/a/1206856).

c2ed8e7